src/Templates/SingleProductTemplateCompatibility.php

```diff
@@ -0,0 +1,425 @@
+ *
+ * @internal
+ */
+class SingleProductTemplateCompatibility extends AbstractTemplateCompatibility {
+	const IS_FIRST_BLOCK = '__wooCommerceIsFirstBlock';
+	const IS_LAST_BLOCK  = '__wooCommerceIsLastBlock';
+
+
+	/**
+	 * Inject hooks to rendered content of corresponding blocks.
+	 *
+	 * @param mixed $block_content The rendered block content.
+	 * @param mixed $block         The parsed block data.
+	 * @return string
+	 */
+	public function inject_hooks( $block_content, $block ) {
+		if ( ! is_product() ) {
+			return $block_content;
+		}
+
+		$this->remove_default_hooks();
+
+		$first_or_last_block_content = $this->inject_hook_to_first_and_last_blocks( $block_content, $block );
+
+		if ( isset( $first_or_last_block_content ) ) {
+			return $first_or_last_block_content;
+		}
+
+		$block_name = $block['blockName'];
+
+		$block_hooks = array_filter(
+			$this->hook_data,
+			function( $hook ) use ( $block_name ) {
+				return $hook['block_name'] === $block_name;
+			}
+		);
+
+		return sprintf(
+			'%1$s%2$s%3$s',
+			$this->get_hooks_buffer( $block_hooks, 'before' ),
+			$block_content,
+			$this->get_hooks_buffer( $block_hooks, 'after' )
+		);
+	}
+
+	/**
+	 * Inject custom hooks to the first and last blocks.
+	 * Since that there is a custom logic for the first and last block, we have to inject the hooks manually.
+	 * The first block supports the following hooks:
+	 * woocommerce_before_single_product
+	 * woocommerce_before_single_product_summary
+	 * woocommerce_single_product_summary
+	 *
+	 * The last block supports the following hooks:
+	 * woocommerce_after_single_product
+	 *
+	 * @param mixed $block_content The rendered block content.
+	 * @param mixed $block         The parsed block data.
+	 * @return string
+	 */
+	private function inject_hook_to_first_and_last_blocks( $block_content, $block ) {
+		$first_block_hook = array(
+			'before' => array(
+				'woocommerce_before_main_content'    => $this->hook_data['woocommerce_before_main_content'],
+				'woocommerce_before_single_product'  => $this->hook_data['woocommerce_before_single_product'],
+				'woocommerce_before_single_product_summary' => $this->hook_data['woocommerce_before_single_product_summary'],
+				'woocommerce_single_product_summary' => $this->hook_data['woocommerce_single_product_summary'],
+			),
+			'after'  => array(),
+		);
+
+		$last_block_hook = array(
+			'before' => array(),
+			'after'  => array(
+				'woocommerce_after_single_product' => $this->hook_data['woocommerce_after_single_product'],
+				'woocommerce_after_main_content'   => $this->hook_data['woocommerce_after_main_content'],
+				'woocommerce_sidebar'              => $this->hook_data['woocommerce_sidebar'],
+			),
+		);
+
+		if ( isset( $block['attrs'][ self::IS_FIRST_BLOCK ] ) && isset( $block['attrs'][ self::IS_LAST_BLOCK ] ) ) {
+			return sprintf(
+				'%1$s%2$s%3$s',
+				$this->get_hooks_buffer(
+					$first_block_hook['before'],
```

Comment:
Should we merge the first and last block's `before` hooks? You are doing so with the [`after` ones](https://github.com/woocommerce/woocommerce-blocks/pull/8442/files#diff-bbdd2400423027ed235c89d88361c9e5e2a4a6e4d28e6282b97a67d406dcdbd8R96-R102).
```suggestion
					array_merge(
						$first_block_hook['before'],
						$last_block_hook['before']
					),
```

Reply:
I will updated it, but `$last_block_hook['before']` is empty.